Optimizing and Reorganizing Quickstart guide (#236)
confidential-containers/confidential-containers: quickstart.md

kubectl apply -k github.com/confidential-containers/operator/config/release?ref=<RELEASE_VERSION>
```

> For example, to deploy the `v0.9.0` release run:

Reviewer: Should we update these versions to `v0.10.0` as that's the lowest version that matches these instructions with the `kata-qemu-coco-dev` runtimeclass?

Author: Updated

```
Output:
```

> NAME                 HANDLER              AGE

Reviewer 2: @mythi should there be enclave-cc here?

Reviewer 3: Good point. I think it would make sense since `enclave-cc.md` does not cover it and it can also co-exist with `kata-qemu-tdx`.

kata                 kata-qemu            8d
kata-clh             kata-clh             8d
kata-qemu            kata-qemu            8d
kata-qemu-coco-dev   kata-qemu-coco-dev   8d
kata-qemu-sev        kata-qemu-sev        8d
kata-qemu-snp        kata-qemu-snp        8d
kata-qemu-tdx        kata-qemu-tdx        8d

```

Details on each of the runtime classes:

> - *kata* - standard kata runtime providing a link to kata-qemu

Reviewer: @fidencio can correct me if I'm wrong, but I think there is a bit of nuance here in that `kata` points to whatever is the default-shim. The default default-shim is kata-qemu, but I'm not sure if we expose the ability to update this in the CoCo operator? 

It might be that it isn't worth going into in this quickstart guide, but we could just reference the handler here and say something like
"Convenience runtime that uses the handler of the default runtime" to cover us?

Author: Updated

> - *kata-clh* - standard kata runtime using the cloud hypervisor including all CoCo building blocks for a non CC HW

Reviewer: I don't think kata-clh has confidential support?

Output:
```
NAME            HANDLER         AGE
kata            kata-qemu       9m55s
kata-clh        kata-clh        9m55s
kata-qemu       kata-qemu       9m55s
kata-qemu-sev   kata-qemu-sev   9m55s
kata-qemu-snp   kata-qemu-snp   9m55s

> kata-qemu-tdx   kata-qemu-tdx   9m55s

Reviewer 4: It's missing kata-coco-qemu-dev.

Author: updated

```

Details on each of the runtime classes:

> - *kata* - standard kata runtime using the QEMU hypervisor including all CoCo building blocks for a non CC HW

Reviewer 4: This has to be modified.

kata provides only a "link" for kata-qemu, but this is not using all CoCo building blocks for a  non TEE HW.
That's on the `kata-coco-qemu-dev` one.

Author: updated

- *kata-clh* - standard kata runtime using the cloud hypervisor including all CoCo building blocks for a non CC HW
- *kata-qemu* - same as kata
- *kata-qemu-sev* - using QEMU, and support for AMD SEV HW
- *kata-qemu-snp* - using QEMU, and support for AMD SNP HW

> - *kata-qemu-tdx* - using QEMU, with TDVF, and support for Intel TDX CC HW, prepared for using Verdictd and EAA KBC.

Reviewer 2: Is this still accurate? @fidencio 
```
... for using Verdictd and EAA KBC.
```

Reviewer 4: No, not accurate anymore.

I'd go for something as simple as: using QEMU, and support Intel TDX HW based on what's provided by Ubuntu (https://github.com/canonical/tdx) and CentOS 9 Stream (https://sigs.centos.org/virt/tdx/).

Author: Updated

- *kata-qemu* - same as kata
- *kata-qemu-tdx* - using QEMU, with TDVF, and support for Intel TDX CC HW, prepared for using Verdictd and EAA KBC.
- *kata-qemu-sev* - using QEMU, and support for AMD SEV HW

> - *kata-qemu-snp* - using QEMU, and support for AMD SNP HW

Reviewer 5: These runtime class names aren't up-to-date with the latest operator. If you can install the operator, you should get the new set of runtime classes and update them.

Author: Updated

enclave-cc      enclave-cc      9m55s
```

The CoCo operator environment has been setup and deployed!

### Platform Setup

While the operator deploys all the required binaries and artifacts and sets up runtime classes that use them,
certain platforms may require additional configuration to enable confidential computing. For example, a specific 
host kernel or firmware may be required. See the [guides](./guides/) for more information.

## Using CoCo

Below is a brief summary and description of some of the CoCo use cases:

- **Container Launch with Only Memory Encryption (No Attestation)** - Launch a container with memory encryption 
- **Container Launch with Encrypted Image** - Launch an encrypted container by proving the workload is running 
  in a TEE in order to retrieve the decryption key
- **Container Launch with Image Signature Verification** - Launch a container and verify the authenticity and 
  integrity of an image by proving the workload is running in a TEE
- **Sealed secret** - Implement wrapped kubernetes secrets that are confidential to the workload owner and are
  automatically decrypted by proving the workload is running in a TEE

> The CoCo use cases are implemented differently for each TEE, and they are described in each corresponding 

Reviewer 2: ```
The CoCo use cases are implemented differently for each TEE, and they are described in each corresponding 
```
Is that really where we've ended up at? 
The existing quickstart (and now the CoCo-dev.md guide) has a generic encrypted container yaml file where you paste in which runtime class you need.

Author: This line is more of a comment to remind the users that each TEE environment requires different steps that they should confirm from the specific TEE guide.

> [guide](./guides) section. To get started using CoCo without TEE hardware, follow the Non-TEE guide below:

Reviewer 2: ```suggestion
[guide](./guides) section. To get started using CoCo without TEE hardware, follow the CoCo-dev guide below:
```

- [CoCo-dev](./guides/coco-dev.md)
- [SEV(-ES)](./guides/sev.md)

> - [SNP](./guides/snp.md)

Reviewer 2: I gather this comes from #237?

> - TDX

Reviewer 2: If you are suggesting this path of abandoning the generic encrypted container yaml then are you creating a TDX guide?

Author: Intel folks are working on the TDX documentation that I will link here when done.

Reviewer 2: Do you know who that might be?

Reviewer 3: > Intel folks are working on the TDX documentation that I will link here when done.

we are not working on it. I'm not convinced we need TEE specific docs unless there's something very special about them that differs from `quickstart.md` flows.

certain platforms may require additional configuration to enable confidential computing. For example, a specific 
host kernel or firmware may be required. See the [guides](./guides/) for more information.

## Using CoCo

Below is a brief summary and description of some of the CoCo use cases:

- **Container Launch with Only Memory Encryption (No Attestation)** - Launch a container with memory encryption 
- **Container Launch with Encrypted Image** - Launch an encrypted container by proving the workload is running 
  in a TEE in order to retrieve the decryption key
- **Container Launch with Image Signature Verification** - Launch a container and verify the authenticity and 
  integrity of an image by proving the workload is running in a TEE
- **Sealed secret** - Implement wrapped kubernetes secrets that are confidential to the workload owner and are
  automatically decrypted by proving the workload is running in a TEE

The CoCo use cases are implemented differently for each TEE, and they are described in each corresponding 
[guide](./guides) section. To get started using CoCo without TEE hardware, follow the Non-TEE guide below:

> - [Non-TEE](./guides/non-tee.md)

Reviewer 5: I think we refer to this as `coco-dev` now. We should probably update that. Also I think we could keep this stuff on the main guide, but I guess it's ok to have a separate page.

Author: Updated

- [SEV(-ES)](./guides/sev.md)
- [SNP](./guides/snp.md)
- TDX
- SGX

> - [Secure Enclave](./guides/enclave-cc.md)

Author: Do you mean we don't need the Secure Enclave bullet because that link is the same as the SGX?

> - ...

Author: Yes definitely, I've added the link.

## Using CoCo

Below is a brief summary and description of some of the CoCo use cases:

- **Container Launch with Only Memory Encryption (No Attestation)** - Launch a container with memory encryption 
- **Container Launch with Encrypted Image** - Launch an encrypted container by proving the workload is running 
  in a TEE in order to retrieve the decryption key
- **Container Launch with Image Signature Verification** - Launch a container and verify the authenticity and 
  integrity of an image by proving the workload is running in a TEE
- **Sealed secret** - Implement wrapped kubernetes secrets that are confidential to the workload owner and are
  automatically decrypted by proving the workload is running in a TEE

The CoCo use cases are implemented differently for each TEE, and they are described in each corresponding 
[guide](./guides) section. To get started using CoCo without TEE hardware, follow the Non-TEE guide below:

- [CoCo-dev](./guides/coco-dev.md)
- [SEV(-ES)](./guides/sev.md)
- [SNP](./guides/snp.md)
- TDX
- [SGX](./guides/enclave-cc.md)
- Secure Enclave
- [IBM Secure Execution](./guides/ibm-se.md)
- ...

> ## Additonal CoCo Features

Reviewer: ```suggestion
## Additional CoCo Features
```

> Here are some of the additional features CoCo supports:

Reviewer 6: We were trying to distinguish major orchestration use cases / flows from supplemental or complimentary features. If you think there's a better way to represent this, we can do that. Or just combine these with the use cases. What you think @stevenhorsman?

> Here are some of the additional features CoCo supports:

Reviewer: I'm just not sure I understand why some of them are major use cases/flows and others are supplemental? Where did that categorisation come from e.g. sealed secrets is a features and secure storage is a feature, what's the difference?

> Here are some of the additional features CoCo supports:

Reviewer 6: Like I mentioned, was trying to distinguish orchestration use case (k8s sealed secret) from a feature that may be complimentary but not required like secure storage. Some of this may be coming from our limited understanding of these additional items. We're just trying to provide a framework to make things more clear in the documentation. Steve, you're prob more familiar with all the features, what makes sense here?

### Platform Setup

While the operator deploys all the required binaries and artifacts and sets up runtime classes that use them,
certain platforms may require additional configuration to enable confidential computing. For example, a specific 
host kernel or firmware may need to be utilized. The guides section has further instructions on how to apply
individual TEE environments.

## Using CoCo

> The CoCo operator environment has been deployed and setup!

Reviewer 7: ```suggestion
The CoCo operator environment has been setup and deployed!
```

  integrity of an image by proving the workload is running in a TEE
- **Sealed secret** - Implement wrapped kubernetes secrets that are confidential to the workload owner and are
  automatically decrypted by proving the workload is running in a TEE

The CoCo use cases are implemented differently for each TEE, and they are described in each corresponding 
[guide](./guides) section. To get started using CoCo without TEE hardware, follow the Non-TEE guide below:

- [Non-TEE](./guides/non-tee.md)
- [SEV(-ES)](./guides/sev.md)
- [SNP](./guides/snp.md)
- TDX
- SGX
- [Secure Enclave](./guides/enclave-cc.md)
- ...

## Additonal CoCo Features

CoCo has full featured support for complete secure cloud virtualized environment. Here are some of the additional 
features it supports:

- **Ephemeral Storage** - temporary storage that is used during the lifecycle of the container but is cleared out 
  when a pod is restarted or finishes its task
- **Authenticated Registries** - secure container registries that require authentication to access and manage container 
  images that ensures that only trusted images are deployed in the Confidential Container

> - **Secure Storage** - mechanisms and technologies used to protect data at rest, ensuring that sensitive information 

Author: Link to doc in kata containers

  automatically decrypted by proving the workload is running in a TEE
- **Ephemeral Storage** - Temporary storage that is used during the lifecycle of the container but is cleared out 
  when a pod is restarted or finishes its task. At the moment, only ephemeral storage of the container itself is 
  supported and it has to be explicityly configured.
- **Authenticated Registries** - Create secure container registries that require authentication to access and manage container 
  images that ensures that only trusted images are deployed in the Confidential Container. The host must have access 
  to the registry credentials.
- **Secure Storage** - Mechanisms and technologies used to protect data at rest, ensuring that sensitive information 
  remains confidential and tamper-proof.
- **Peer Pods** - Enable the creation of VMs on any environment without requiring bare metal servers or nested 
  virtualization support. More information about this feature can be found [here](https://github.com/confidential-containers/cloud-api-adaptor/tree/main).
  

  The CoCo use cases are implemented differently for each TEE, and they are described in each corresponding 

> [guide](./guides) section. To get started using CoCo without TEE hardware, follow the CoCo-dev guide below:

Reviewer 5: This section blends in with the stuff above it. Can you add a heading for `Platforms`

Author: Updated

```

Now go back to the k8s node and ensure that you still don’t have any bitnami/nginx images on it:
```
root@cluster01-master-0:/home/ubuntu# crictl  -r  unix:///run/containerd/containerd.sock image ls | grep bitnami/nginx
```

## Encrypted and/or signed images with attestation

The previous example does not involve any attestation because the workload container isn't signed or encrypted
and the workload itself does not require any secrets.

This is not the case for most real workloads. It is recommended to use CoCo with signed and/or encrypted images.
The workload itself can also request secrets from the attestation agent in the guest.

Secrets are provisioned to the guest in conjunction with an attestation, which is based on hardware evidence.
The rest of this guide focuses on setting up more substantial encrypted/signed workloads using attestation
and confidential hardware.

CoCo has a modular attestation interface and there are a few options for attestation.
CoCo provides a generic Key Broker Service (KBS) that the rest of this guide will be focused on.
The SEV runtime class uses `simple-kbs`, which is described in the [SEV guide](./guides/sev.md).

> There is also `eaa_kbc`/`verdictd` which is described [here](./guides/eaa_verdictd.md).

Reviewer 3: `./guides/eaa_verdictd.md`  could be `git rm`d in this PR. EAA KBC was recently dropped in AA.

>   The CoCo use cases are implemented differently for each TEE, and they are described in each corresponding 
> [guide](./guides) section. To get started using CoCo without TEE hardware, follow the CoCo-dev guide below:

Reviewer 3: Is the purpose here to list CoCo platforms or give pointers to docs where the setup/use-cases differ:
```suggestion
  With some TEEs, the CoCo use cases and/or configurations are implemented differently. Those are described in each corresponding 
[guide](./guides) section. To get started using CoCo without TEE hardware, follow the CoCo-dev guide below:
```

Reviewer 3: My main concern is that there's very little TEE specifics in the new sub-docs that are being created so we just copy-paste generic stuff there and create maintenance overhead. "how to setup the host" links can be added here in the `quickstart.md` similar to how it's done for TDX right now.

Author: Do you want to do this in this PR? My original intention was to deal with sub doc reorganization in the next PR, but if you feel it needs to be updated now, we can address it now too.

Reviewer 3: I think next PR is fine.

> - TDX

Reviewer 2: So should we make this clear like this ...
```suggestion
- TDX: No additional steps required.
```

> - Secure Enclave

Reviewer 2: I think Secure Enclave either meant IBM Secure Execution or Intel SGX. Either way I think this bullet goes away unless I'm forgetting someone's TEE?

```suggestion
```

Author: yes, I believe it was SGX, just removed and updated.
